src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -56,15 +56,21 @@
          (jdbc/with-query-results-cursor results-query (comp row-fn munge-fn)))))))
 
 (defn produce-streaming-body
-  "Given a query, and database connection, return a Ring response with the query
-  results.
+  "DEPRECATED - this function will be replaced by produce-streaming-body'
+                which accepts a query map
+
+  Given a query, and database connection, return a Ring response with
+  the query results.
 
   If the query can't be parsed, a 400 is returned."
+  {:deprecated "3.0.0"}
   [entity version query paging-options db url-prefix]
   (try
     (jdbc/with-transacted-connection db
       (let [[query->sql munge-fn] (entity->sql-fns entity version paging-options url-prefix)
-            {:keys [results-query count-query]} (-> query (json/parse-strict-string true) query->sql)
+            {:keys [results-query count-query]} (query->sql (if (string? query)
```

Comment:
Another odd area. If we have a JSON doc as the body of a POST, the query has already been parsed. In other areas of the code, we have emitted the json, so that it can be re-parsed here.